// fpm/lib/commands/index.js
"use strict";

module.exports = {
    project: require('./project'),     //创建一个项目
    install: require('./install'),     //组件安装
    register : require('./register'),  //组件注册
    info:   require('./info'),         //组件版本信息
    uninstall : require('./uninstall'), //卸载一个组件
    update : require('./update'), //更新一个组件
    commit : require('./commit'), //更新一个组件
    list : require('./list'), //更新一个组件
    module : require('./module'), //创建组件项目
    build : require('./build'),   //build项目
    replacecdn : require('./replacecdn'), //替换cdn
    uploadcdn : require('./uploadcdn'),   //上传文件cdn
    reg : require('./register'),
    search : require('./search'),
    debug : require('./debug'),
    server : require('./server'),
    chver : require('./chver'), //改变js/css版本
    create: require('./create'), // 创建一个git项目
    initconfig: require('./initconfig'), // 填写初始信息
    clone: require('./clone'), // 从git上clone一个项目
    installenv: require('./installenv'), // 安装项目的支持环境    
    publish: require('./publish')
};
